### DISCUSSION TITLE
Replies from Marlin not displayed

### DISCUSSION THREAD

**Discussion by gerthalberg** [luc-github/ESP3D#589] [Questions]
I've put the ESP3D on an ESP32 Devkit, and have set the RX/TX pins to 40/41. The page is up and running. Connected the ESP to a BTT TFT35_E3_V3.0 at the WiFi pins. The TFT works normal in both touch and marlin mode, The ESP3D can send commands to marlin. I can set heater temps and start heating. I can move the printhead and home it But I don't get any data returned - so I can't see the current position and I can't see the current temp. Can someone pls. tell me what I've done wrong.

**Comment by luc-github** [comment]
why change rx/tx pins?

**Reply by gerthalberg** [reply]
Well changing the pins was the way I could find - that would send data to the printer. I just tried to reflash using the default pins i.e. -1 and -1. That allows me to receive data from the printer but I can not send data to the printer. So if I use the default I can receive if I use 40/41 I can send but not at the same time. As a side item - when I use the default pins, the TFT touch will not send anything to the printer I need to switch to Marlin mode in order to control the printer. In Marlin I have the serial defined as SERIAL_PORT -1 and SERIAL_PORT2 0

**Reply by luc-github** [reply]
1 - there is is no pins 40/41 you must use GPIO number which are 1 and 3 which are default ones, same as set -1 
2 - how do you connect / wire your esp32 to TFT ?
3 - What FW version of TFT are you using ?

**Reply by gerthalberg** [reply]
Thank's Luc - that probably explains a lot - but as I write I tried with the default and although I did receive data I was not able to send anything as opposed to compiling it with the wrong numbers. To connect the ESP to the TFT I made a little harness with Dupont connectors 2X4 at the tft end and single at the ESP end. I run 3.3v and GND from the WiFi socket to the ESP and also from the RX/TX on the WiFi plug to the u0RXD/u0TXD - which are GPIO1/3. The TFT firmware is Version 0.26.X compiled Jan 18 after a refresh from BTT Github.
So if I want to use other pins I should use the GPIO number is that right? e.g to use the U2RXD/TXD it would be GPIO16/17

**Reply by luc-github** [reply]
I would say do not try to change pins - it is not the solution and I do not know if you do another change in code than using wring pins but I really suggest you to stick to unmodified FW first - it should work out of the box

instead of mess up the esp3d fw let's try to isolate the root cause - the fw is not the issue and changing pins is not the way to go unless your esp boad has faulty pins which is very very low probability
1 - do a clean install without any modification in FW and share the `[ESP420]plain` output
2 - let check your hardware first : do you have a ftdi or serial USB adapter to check the esp serial pins output ?
3 - when using rx/tx you cannot use usb on dev kit - can you confirm USB on esp is not connected ?
4 - can you do `M115 TFT` in webUI terminal when ESP is connected to TFT
5 - try to send `M105` to see if the temperatures are show

**Comment by gerthalberg** [comment]
Luc,

Thank’s a million for shaking me out of my little tantrum. I did as suggested and after a few tries it’s now up and running. for information the following is a transcript of what I did.

Transcript preparing ESP32 Devkit

Erasing board
Building in release mode
Looking for serial port...
Auto-detected: /dev/cu.usbserial-0001
Erasing...
esptool.py v3.0
Serial port /dev/cu.usbserial-0001
Connecting.....
Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: ac:67:b2:2c:8d:20
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Chip erase completed successfully in 3.3s
Hard resetting via RTS pin...

Build File System Image
Building in release mode
Building SPIFFS image from 'esp3d/data' directory to .pioenvs/esp32dev/spiffs.bin
/favicon.ico
/404.htm
/index.html.gz

Build
Archiving .pioenvs/esp32dev/libFrameworkArduino.a
Indexing .pioenvs/esp32dev/libFrameworkArduino.a
Linking .pioenvs/esp32dev/firmware.elf
Retrieving maximum program size .pioenvs/esp32dev/firmware.elf
Checking size .pioenvs/esp32dev/firmware.elf
Building .pioenvs/esp32dev/firmware.bin
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]  13.4% (used 43904 bytes from 327680 bytes)
Flash: [=====     ]  54.5% (used 1071942 bytes from 1966080 bytes)
esptool.py v3.0

Upload
Wrote 1072064 bytes (618323 compressed) at 0x00010000 in 8.6 seconds (effective 999.5 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

Upload File System Image
Wrote 196608 bytes (115416 compressed) at 0x003d0000 in 1.7 seconds (effective 950.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

Configuring ESP3D



Output from [ESP420]plain
[ESP420]plain
Chip ID: 8333
CPU Frequency: 240Mhz
CPU Temperature: 91.7C
Free memory: 229.73 KB
SDK: v3.2.3-14-gd3e562907
Flash Size: 4.00 MB
Available Size for update: 1.87 MB(Ok)
Available Size for SPIFFS: 169.38 KB
Baud rate: 250000
Sleep mode: Light
Channel: 11
Phy Mode: 11g
Web port: 80
Data port: 8888
Hostname: ESPCUBE
Active Mode: STA (AC:67:B2:2C:8D:20)
Connected to: WiFimodem-ABBE
Signal: 98%
IP Mode: DHCP
IP: 192.168.0.21
Gateway: 192.168.0.1
Mask: 255.255.255.0
DNS: 212.242.40.3
Disabled Mode: AP (AC:67:B2:2C:8D:21)
Captive portal: Enabled
SSDP: Enabled
NetBios: Disabled
mDNS: Enabled
Web Update: Enabled
Pin Recovery: Disabled
Authentication: Disabled
Target Firmware: Marlin
M117 output: Enabled
Notifications: Disabled
Serial output: Enabled
Web socket  output: Enabled
TCP output: Enabled
FW version: 2.1.1.b10 ESP32

M115 Output
M115
FIRMWARE_NAME:Marlin 2.0.7.2 (Feb 13 2021 14:56:09) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Gert's Cube EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff
Cap:SERIAL_XON_XOFF:0
Cap:BINARY_FILE_TRANSFER:1
Cap:EEPROM:1
Cap:VOLUMETRIC:1
Cap:AUTOREPORT_TEMP:1
Cap:PROGRESS:0
Cap:PRINT_JOB:1
Cap:AUTOLEVEL:1
Cap:RUNOUT:1
Cap:Z_PROBE:1
Cap:LEVELING_DATA:1
Cap:BUILD_PERCENT:0
Cap:SOFTWARE_POWER:0
Cap:TOGGLE_LIGHTS:0
Cap:CASE_LIGHT_BRIGHTNESS:0
Cap:EMERGENCY_PARSER:1
Cap:PROMPT_SUPPORT:1
Cap:SDCARD:1
Cap:AUTOREPORT_SD_STATUS:1
Cap:LONG_FILENAME:1
Cap:THERMAL_PROTECTION:1
Cap:MOTION_MODES:0
Cap:ARCS:1
Cap:BABYSTEPPING:1
Cap:CHAMBER_TEMPERATURE:0
area:{full:{min:{x:0.0000,y:0.0000,z:0.0000},max:{x:420.0000,y:400.0000,z:400.0000}},work:{min:{x:0.0000,y:0.0000,z:0.0000},max:{x:420.0000,y:400.0000,z:400.0000}}}
ok P31 B3

And finally



Thank’s for keeping calm.

Regards

Gert Halberg

**Reply by luc-github** [reply]
Ok esp3d installation seems ok 👍 
* please do `M115 TFT` not just `M115`
* also what about `M105` ?
* what is webUI version ?

**Comment by gerthalberg** [comment]
Output from M115 TFT
 T:240.31 /240.0000 B:37.83 /75.0000 @:10 B@:127
 T:240.31 /240.0000 B:38.08 /75.0000 @:0 B@:127
 T:240.00 /240.0000 B:38.50 /75.0000 @:38 B@:127
 T:240.31 /240.0000 B:38.75 /75.0000 @:44 B@:127
M115 TFT
FIRMWARE_NAME: BIGTREE_TFT35_V3.0_E3.26.x SOURCE_CODE_URL:https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware
Cap:HOTEND_NUM:1
Cap:EXTRUDER_NUM:1
Cap:FAN_NUM:1
Cap:FAN_CTRL_NUM:1
ok
output from M105
M105
ok T:240.31 /240.0000 B:47.11 /75.0000 @:33 B@:127
However - when I try to preheat from ESP3D and the TFT is in touch mode It gives a checksum error but if I switch the TFT to Marlin mode I can start to heat from ESP3D

/Gert

**Reply by luc-github** [reply]
so looks like now communication is working ^_^
The checksum issue seems TFT Firmware issue - I think some people already raised the issue in BTT issue tracker

**Comment by gerthalberg** [comment]
Sounds very plausible - I’m contemplating, if I should in fact move the ESP to the board plug instead that would circumvent the TFT error, I have an SKR 1.4 Turbo . As far as I can understand I need to switch to the serial port that’s connected to the WiFi plug and thereby disabling the USB - which doesn’t matter as I don’t use it.
Anyway it’s up and running and able to do pretty much what I wanted - so I’m happy.

Regards.

Gert

**Comment by gerritv** [comment]
I recently went through the experience of connecting ESP3D to an SKR 1.4 Turbo.
USB connection still works with ESP01s on #define SERIAL_PORT_2 3, leave #define SERIAL_PORT -1 for USB connection

You can use both ESP3D and Octoprint/Repetier at same time.

**Reply by luc-github** [reply]
The issue is Marlin allows only 2 outputs
1 - USB
2 - serial WIFI or serial TFT 

* if you use TFT (touch mode) and want wifi you only can plug esp to TFT,
* if you use Screen in GLCD emulation mode only (no touch) then you may connect esp board on printer board directly

**Comment by gerthalberg** [comment]
Thank’s Gerrit - but does the TFT Touch still work as that also requires a serial connection - It doesn’t bother me if the USB is disconnected I’m not using it anyway. As I wrote initially I really only want to be able to start heating and then monitor temp and height. I’m always by the printer when I start a print.

Regards

Gert

**Reply by gerritv** [reply]
Ah, not sure, I don't have the TFT. However I would be optimistic. The schematic shows WiFi on RXD3/TXD3 while the TFT is on RXD0/TXD0 so totally separate ports.

Are you certain that the TFT is serial? In Configuration.h it shows SPI or I2C, not serial.

**Comment by gerthalberg** [comment]
Yes I suspected that much - but if I don’t care about the USB since I don’t use it wouldn’t it be possible to say e.g.
1 - port 2 or 3 for the WiFi
2 SERIAL2 0

Regards

Gert
